libimage/manifests/manifests.go

```diff
@@ -140,41 +184,79 @@ func (l *list) SaveToImage(store storage.Store, imageID string, names []string,
 	if err != nil {
 		return "", err
 	}
-	img, err := store.CreateImage(imageID, names, "", "", &storage.ImageOptions{})
-	if err == nil || errors.Is(err, storage.ErrDuplicateID) {
-		created := (err == nil)
-		if created {
-			imageID = img.ID
-			l.instances[""] = img.ID
-		}
-		err := store.SetImageBigData(imageID, storage.ImageDigestManifestBigDataNamePrefix, manifestBytes, manifest.Digest)
-		if err != nil {
-			if created {
-				if _, err2 := store.DeleteImage(img.ID, true); err2 != nil {
-					logrus.Errorf("Deleting image %q after failing to save manifest for it", img.ID)
+	artifactsBytes, err := json.Marshal(&l.artifacts)
+	if err != nil {
+		return "", err
+	}
+	manifestDigest, err := manifest.Digest(manifestBytes)
+	if err != nil {
+		return "", err
+	}
+	imageOptions := &storage.ImageOptions{
+		BigData: []storage.ImageBigDataOption{
+			{Key: storage.ImageDigestManifestBigDataNamePrefix, Data: manifestBytes, Digest: manifestDigest},
+			{Key: instancesData, Data: instancesBytes},
+			{Key: artifactsData, Data: artifactsBytes},
+		},
+	}
+	img, err := store.CreateImage(imageID, names, "", "", imageOptions)
+	if err != nil {
+		if imageID != "" && errors.Is(err, storage.ErrDuplicateID) {
+			for _, bd := range imageOptions.BigData {
+				digester := manifest.Digest
+				if !strings.HasPrefix(bd.Key, storage.ImageDigestManifestBigDataNamePrefix) {
+					digester = nil
+				}
+				err := store.SetImageBigData(imageID, bd.Key, bd.Data, digester)
+				if err != nil {
+					return "", fmt.Errorf("saving manifest list to image %q: %w", imageID, err)
 				}
 			}
-			return "", fmt.Errorf("saving manifest list to image %q: %w", imageID, err)
+			return imageID, nil
 		}
-		err = store.SetImageBigData(imageID, instancesData, instancesBytes, nil)
-		if err != nil {
-			if created {
-				if _, err2 := store.DeleteImage(img.ID, true); err2 != nil {
-					logrus.Errorf("Deleting image %q after failing to save instance locations for it", img.ID)
-				}
+		return "", err
+	}
+	l.instances[""] = img.ID
+	return img.ID, nil
+}
+
+// Files returns the list of files associated with a particular artifact
+// instance in the image index, primarily for display purposes.
+func (l *list) Files(instanceDigest digest.Digest) ([]string, error) {
+	filesList, ok := l.artifacts.Files[instanceDigest]
+	if ok {
+		return append([]string{}, filesList...), nil
+	}
+	return nil, nil
+}
+
+// instanceByFile returns the instanceDigest of the first manifest in the index
+// which refers to the named file.  The name will be passed to filepath.Abs()
+// before searching for an instance which references it.
+func (l *list) InstanceByFile(file string) (digest.Digest, error) {
+	if parsedDigest, err := digest.Parse(file); err == nil {
```

Comment:
Is it forbidden to use files named like that? (I am generally worried about strings that have different semantics purely based on contents, and) If so, it should be forbidden when adding the file.

Reply:
I’m not sure why digest-looking file names are rejected at all; is that because some heuristic string parsing anticipated to happen at the CLI level?

Either way, the immediate consequence I am worried about is that `AddArtifact(…., []string{"sha256:aaa…"})`, is AFAICS, currently accepted without protest, but a follow-up `InstanceByFile("sha256:aaa…")` refuses to look for the file. That seems surprising.

Reply:
My expectation is that remembering (or copying and pasting) digests for repeated calls to `buildah manifest` is tiresome.  I think that being able to describe which descriptor in the image index we want to modify by naming the thing that it refers might be easier.
That doesn't work as well if another index's contents were added using `buildah manifest add --all`, since that doesn't create a reference to the other index so much as it copies that index's descriptor list into the index being modified, and we don't get information about what those descriptors refer to that we could stash in the `instances` list to use this way.

Reply:
> I think that being able to describe which descriptor in the image index we want to modify by naming the thing that it refers might be easier.

That makes sense to me.

I don’t immediately see how that relates to this code refusing to look for digest-named files. Is that because the CLI would try to parse the same CLI argument both ways? Even if that were the case, the way I think about it, either this is a CLI-only matter (the CLI can have that two-way heuristic, and the CLI can refuse to create digest-named files), or it is a mater explicitly supported by this lower layer, and then not only should `InstanceByFile` refuse to search, but `AddArtifact` should refuse to accept digest-named files.

Or do you mean that `buildah manifest add --all` would create digest-named files, so they ~need to be allowed?

Reply:
Yeah, the related changes to the CLIs would hand the strings args they currently get to this method instead of only accepting digest values directly.

The `manifest add --all` bit was a digression, mainly about how this doesn't provide any help for that case, since we don't have an image name or file name to match an incoming value against, so those instances would still have to be specified by digest.

I'm not completely sold on rejecting file names that look like digests outright, since there's no ambiguity when we're first adding them, but it does seem like it would be a rare case.

Reply:
> we don't have an image name or file name to match an incoming value against

That `.title` annotation mentioned elsewhere in this PR might work.

---

If the `./…` approach works for searching, I can live with the code as is… Aesthetically I’m not a fan, but, \*shrug\*